app/controllers/admin/vouchers_controller.rb

```diff
@@ -9,9 +9,13 @@ def new
     end
 
     def create
+      # In the scenario where you get an error when trying to create a percentage voucher, we'll
+      # now have percentage rate voucher instanciated. Hence why we check for both params type
+      voucher_type = params.dig(:vouchers_flat_rate, :voucher_type) ||
+                     params[:vouchers_percentage_rate][:voucher_type]
```

Comment:
You do a similar thing in `permitted_resource_params` already. It could be good to re-use that. But I think I just found a better way of doing that. You can do `Voucher.new.becomes(Voucher)` and then we have only one name in the parameters.

And if we assign the type as attribute to `Voucher` then Rails verifies that it's a subclass and we don't have to do that. I'll have a play with it and maybe add a commit. Stay tuned...